File: app/src/main/java/taxi/kassa/ui/adapters/WithdrawsAdapter.java
package taxi.kassa.ui.adapters;

import android.content.Context;
import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import java.util.List;

import taxi.kassa.R;
import taxi.kassa.responses.Withdraw;

public class WithdrawsAdapter extends RecyclerView.Adapter<WithdrawsAdapter.ViewHolder> {

    private LayoutInflater inflater;
    private List<Withdraw> ts_requests;
    private Context context;

    public WithdrawsAdapter(Context context, List<Withdraw> ts_requests) {
        this.ts_requests = ts_requests;
        this.inflater = LayoutInflater.from(context);
        this.context = context;
    }

    @Override
    public WithdrawsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.withdrawal_item, parent, false);
        return new ViewHolder(view);
    }


    @Override
    public void onBindViewHolder(WithdrawsAdapter.ViewHolder holder, int position) {
        final Withdraw info = ts_requests.get(position);
        // date
        String date = info.getDate();
        // source
        int source = Integer.parseInt(info.getSource_id());
        String source_title = "";
        if (source == 2) source_title = "City Такси";
        else if (source == 3) source_title = "Gett Такси";
        else if (source == 1) source_title = "Яндекс.Такси";
        // amount
        String amount = info.getAmount() + " \u20BD";
        // status
        String status = info.getStatus();
        int status_id = info.getStatusId();
        // info
        holder.ts_source.setText(source_title);
        holder.ts_amount.setText(amount);
        holder.ts_date.setText(date);
        holder.ts_status.setText(status);
        // color
        if (status_id == 1 || status_id == 2) holder.ts_status.setTextColor(Color.parseColor("#00a651"));
        else if (status_id == -1) holder.ts_status.setTextColor(Color.parseColor("#cf2832"));
        else holder.ts_status.setTextColor(Color.parseColor("#0f2f5d"));
    }

    @Override
    public int getItemCount() {
        return ts_requests.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        final TextView ts_source, ts_amount, ts_date, ts_status;
        ViewHolder(View view){
            super(view);
            ts_source = view.findViewById(R.id.ts_source);
            ts_amount = view.findViewById(R.id.ts_amount);
            ts_date = view.findViewById(R.id.ts_date);
            ts_status = view.findViewById(R.id.ts_status);
        }
    }
}

File: app/src/main/java/taxi/kassa/view/withdraws/WithdrawsAdapter.kt
package taxi.kassa.view.withdraws

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.item_withdraw.view.*
import kotlinx.android.synthetic.main.item_withdraw_date.view.*
import taxi.kassa.R
import taxi.kassa.model.responses.Withdraw
import taxi.kassa.util.Constants.APPROVED
import taxi.kassa.util.Constants.DENIED
import taxi.kassa.util.Constants.NEW
import taxi.kassa.util.Constants.WITHDRAWN
import taxi.kassa.util.setFormattedText

class WithdrawsAdapter(
    withdraws: MutableList<Withdraw>,
    private val clickListener: (Withdraw) -> Unit
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val TYPE_ITEM = 0
        private const val TYPE_DATE = 1
        private const val DATE_ITEM_ID = "777"
    }

    private val withdraws = mutableListOf<Withdraw>()

    init {
        this.withdraws.addAll(withdraws)
        val dates = mutableSetOf<Withdraw>()

        if (this.withdraws.isNotEmpty()) {
            var lastDate = this.withdraws[0].getDate()

            dates.add(Withdraw(DATE_ITEM_ID, "0", "", this.withdraws[0].date, ""))

            this.withdraws.map {
                if (it.getDate() != lastDate) {
                    dates.add(Withdraw(DATE_ITEM_ID, "0", "", it.date, ""))
                    lastDate = it.getDate()
                }
            }

            this.withdraws.addAll(dates)
            this.withdraws.sortBy { it.date }
            this.withdraws.reverse()
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            TYPE_ITEM -> {
                WithdrawsViewHolder(
                    LayoutInflater.from(parent.context).inflate(
                        R.layout.item_withdraw,
                        parent,
                        false
                    )
                )
            }
            TYPE_DATE -> {
                DateViewHolder(
                    LayoutInflater.from(parent.context).inflate(
                        R.layout.item_withdraw_date,
                        parent,
                        false
                    )
                )
            }
            else -> WithdrawsViewHolder(
                LayoutInflater.from(parent.context).inflate(
                    R.layout.item_withdraw,
                    parent,
                    false
                )
            )
        }
    }

    override fun getItemCount(): Int = withdraws.size

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val withdraw = withdraws[position]

        when (holder) {
            is WithdrawsViewHolder -> {
                holder.time.text = withdraw.hours
                holder.amount.setFormattedText(R.string.balance_format, withdraw.amount.toDouble())
                holder.status.text = withdraw.status

                when (withdraw.status) {
                    NEW -> holder.statusImage.setImageResource(R.drawable.ic_yellow_circle)
                    APPROVED -> holder.statusImage.setImageResource(R.drawable.ic_yellow_circle)
                    WITHDRAWN -> holder.statusImage.setImageResource(R.drawable.ic_green_circle)
                    DENIED -> holder.statusImage.setImageResource(R.drawable.ic_red_circle)
                }

                when (withdraw.sourceId.toInt()) {
                    1 -> {
                        holder.taxiIcon.setImageResource(R.drawable.ic_yandex_mini)
                        holder.taxiName.setText(R.string.yandex_title)
                    }
                    2 -> {
                        holder.taxiIcon.setImageResource(R.drawable.ic_citymobil_mini)
                        holder.taxiName.setText(R.string.citymobil_title)
                    }
                    3 -> {
                        holder.taxiIcon.setImageResource(R.drawable.ic_gett_mini)
                        holder.taxiName.setText(R.string.gett_title)
                    }
                }

                holder.itemView.setOnClickListener { clickListener(withdraw) }
            }
            is DateViewHolder -> {
                holder.date.text = withdraw.getDate()
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return when (withdraws[position].sourceId) {
            DATE_ITEM_ID -> TYPE_DATE
            else -> TYPE_ITEM
        }
    }

    class WithdrawsViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val time: TextView = view.time_tv
        val taxiIcon: ImageView = view.taxi_icon
        val taxiName: TextView = view.taxi_name
        val amount: TextView = view.amount_tv
        val status: TextView = view.status_tv
        val statusImage: ImageView = view.circle_image
    }

    class DateViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val date: TextView = view.date
    }
}